### DISCUSSION TITLE
Accessing store actions in script

### DISCUSSION THREAD

**Discussion by sylcha** [quasarframework/quasar#12658] [General - Components / Directives / etc] [answered]
Hi there!

I'm struggling to access my store action within the script (not the template) on a Vue file.

I have two stores `names` and `nombre`. There are referenced properly as module in `src/store/index.js`.

In my script part, I import `useStore` from `vuex` and declare a variable to access my store:

```js
const store = useStore();
```

I can access successfully one state property `nombreCourant` of one of my store `nombre` doing:

```js
console.log(store.state.nombre.nombreCourant)
```
Of course, I can change the value of this property from the `template` part, calling one of my actions: in the script part

```js
methods: {
    ...mapActions("nombre", ["changeNombre"]),
  }
```
and in my template part:

```html
<q-btn label="Change" @click="changeNombre({newValue: '12,3'})" />
```

However, I cannot figure out how to change the value of this property **within the script part** because I did not have access to this `changeNombre()` method.

I've try to use `dispatch` but I've surely not used it well because it failed... 

Is there anyone who would explain how to do it to me? Thanks in advance,
Sylvain

**Comment by MilosPaunovic** [comment]
It would be best if you could create a code reproduction for someone to be able to take a deeper dive into it.

**Comment by sylcha** [comment]
OK. I thought it was too much code... but here it is.

## Problem

Be able to change the value of a state property within the script part of a Vue file.

## Organisation of the store

### In `src/store` 

I have the store declare as a module

```js
import { store } from "quasar/wrappers";
import { createStore } from "vuex";
import nombre from "./nombre";

export default store(function (/* { ssrContext } */) {
  const Store = createStore({
    modules: {
      names, // <-- I have a second store declared...
      nombre,
    },

    strict: process.env.DEBUGGING,
  });

  return Store;
});
```

### In `src/store/nombre` 

I have all the states, getters, mutations and actions declared:

**States**

```js
export default function () {
  return {
    nombreCourant: "123.45",
  };
}
```

**Mutations**

```js
export function CHANGE_NOMBRE_COURANT(state, payload) {
  state.nombreCourant = String(payload);
}
```

**Actions**

```js
export function changeNombre({ commit }, payload) {
  commit("CHANGE_NOMBRE_COURANT", payload.newValue);
}
```

**Getters**
```js
export function nbCourant(state) {
  return state.nombreCourant;
}
```
**Index**
```js
import state from './state'
import * as getters from './getters'
import * as mutations from './mutations'
import * as actions from './actions'

export default {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
}
```

## The Vue file

*Full code below*

### In the template

Thanks to `mapActions` and `mapGetters`, I can access and change the value of the state property `nombreCourant`. The value is shown properly and the button update the value correctly.

### In the script

I create a variable `store` using the `useStore()` method. Through this `store` variable, I can access the value of the same property **within the script**

```js
console.log("Nombre courant : " + store.state.nombre.nombreCourant );
```
> What I would like to do (and I cannot figure out how) is to change the value of this `nombreCourant` property **within the script**.

### Vue file

```html
<template>
  <q-dialog position="bottom">
    <q-card>
      <q-toolbar class="bg-secondary text-white" dense>
        <q-toolbar-title>Nouveau nombre</q-toolbar-title>
      </q-toolbar>
      <q-card-section class="text-h5">
        Old value: {{ nbCourant }}
      </q-card-section>
      <q-card-section class="text-h5">
        <q-btn
          v-close-popup
          label="Change"
          @click="changeNombre({ newValue: '333,33' })"
        />
      </q-card-section>
    </q-card>
  </q-dialog>
</template>

<script>
import { mapActions, mapGetters, useStore  } from "vuex";
export default {
  setup() {
    const store = useStore();
    console.log(
          "Nombre courant : " + store.state.nombre.nombreCourant
        );
  },
  computed: {
    ...mapGetters("nombre", ["nbCourant"]),
  },
  methods: {
    ...mapActions("nombre", ["changeNombre"]),
  },
};
</script>
```

**Comment by MilosPaunovic** [comment]
You are mixing Composition API with Options API in your component script - that seems like a first thing to improve.

When I suggested to create a reproduction, what I meant is a Codepen, JSFiddle, Codesandbox or public GitHub repository, for someone to be able to play with it and help you solve issue quicker.

**Reply by sylcha** [reply]
Thanks for your advice. I really new to `Vue` and `Vuex`  (I'm using it for two weeks right now) and I  am not an experienced developper (that's not my job) so I definitely miss some perspective here! I'm aware of the Option and Composition *styles* but I really need some time to get use to it.

I eventually find a solution that I explain below...

**Comment by sylcha** [comment, marked as answer]
Actually, I eventually get the solution. I was incorrectly naming my action. 

## Changing the state of a store using actions in your script

You have to use the `dispatch` function and use it with your `store` object created with `useStore()`. Than, you pass as first argument the name of your action that way: `<name_of_your_store>/<name_of_the_action>`. The second argument is the payload that your action might need.

In my case:

```js
store.dispatch("nombre/changeNombre", {
  newValue: nombreSaisi.value,
});
```
